### DISCUSSION TITLE
Symfony mailer : how to reset "All transports failed." error between two emails sending ?

### DISCUSSION THREAD

**Discussion by sdespont** [symfony/symfony#52530] [Q&A] [answered]
I'm using Symfony Mailer with the `roundroubin` feature containing 4 email server accounts.

I have a use case where a customer needs to send 4 emails in a row from our application. All four emails are sent to a single recipient, one after the other synchronously.

This works well with Symfony Mailer if all the emails can be sent, but if one of the emails cannot be sent, the rest of the emails are not sent with the error message `All transports failed.`.

Example:
1. ...@... <-- valid email 
2. test@example.com <- no MX record, sending of emails failed
3. ...@... <-- valid email <- not sent due to above error
4. ...@... <-- valid email <- not send due to error above

My expected result is that emails 1, 3 and 4 are sent, but email 2 fails.

The validity of all emails is checked before sending with the https://github.com/egulias/EmailValidator library, but as [test@example.com](mailto:test@example.com) is considered valid, the application tries to send the email.

Is there any way of "resetting" the transport failure options between two email sending ?

**Comment by ghost** [comment]
Please always include Component version in use.

> the rest of the emails are not sent with the error message All transports failed..

I don't think this is what happens.
At first I understood your description that emails 3 and 4 are not sent because exception with a message `All transports failed` occurs while sending any of them. However I think that what happens is that the lines with sending emails 3&4 are not even executed because line sending email 2 throws. 
That said this is not a Symfony error. There is nothing to reset. You have to write code that handles this specific case. 
You can also integrate Messenger with Mailer and send messages asynchronously. In this case email 2 would have been returned to the queue and then email 3 would have been successfully sent, and then email 4 also.



However ever if you handle this exception somehow or move to the async handling then RaoundRobinTransport still will cause problem because there is 60 seconds freeze for every transport that fails. This could be made configurable on the Symfony side.

**Comment by sdespont** [comment]
Thank you for your answer, and sorry about the missing details. I am using SF5.4.31.

All the email logic is well encapsulated within a try/catch block with logs that provide insight into the fact that subsequent emails are not sent due to the failure of the 2nd email.

Here is an idea of the code I am using:

```
$recipients = [
    new Address("valid1@gmail.com", "Valid 1"),
    new Address("test@example.com", "Valid email but no MX record for the domain"),
    new Address("valid3@gmail.com", "Valid 1"),
    new Address("valid4@gmail.com", "Valid 1"),
];

$emailFrom = new Address("validexpeditor@gmail.com", "Valid Expeditor");

foreach ($recipients as $recipient) {
    try {
        $this->logger->logNoticeMsg("Send email for contact " . $recipient->getAddress()->getAddress());

        $email = new Email();
        $email->from($emailFrom)
            ->to($recipient)
            ->subject($subject)
            ->html($body);

        $signer = new DkimSigner(
            'file://' . $dkimPath,
            $this->getDomainByEmail($userGroup->getEmailExpeditor()),
            self::DKIM_SELECTOR
        );
        $signedMessage = $signer->sign($email);
        $this->mailerInterface->send($signedMessage);

        $this->logger->logNoticeMsg("Email sent");
    } catch (Exception $e) {
        $this->logger->logErrorMsg("Error when sending email [" . $e->getMessage() . "]");
    }
}
```

The first email is well sent, the others all failed with error log "Error when sending email [No transports found.]". If test@example.com recipient is replaced with another valid email, all the emails are well sent.

Your ideas are very welcome.

**Reply by ghost** [reply, marked as answer]
> The first email is well sent, the others all failed with error log "Error when sending email [No transports found.]".

I bet it's 3rd and 4th email that fail with error log `Error when sending email [No transports found]`." and 2nd fails with `Error when sending email [All transports failed]`.

> Your ideas are very welcome.

Now that we know you have `try..catch` block then this paragraph in my answer applies:
> However ever if you handle this exception somehow or move to the async handling then RaoundRobinTransport still will cause problem because there is 60 seconds freeze for every transport that fails. 

Unfortunatelly it's not configurable at the moment. I've opened an issue for this https://github.com/symfony/symfony/issues/52551

P.S. I don't think the `try..catch` block in you code sample is valid. The secong loop iteration will throw and mails 3 and 4 will not be even tried to be sent. This block should be inside the `foreach` block.

**Comment by sdespont** [comment]
Thanks for your answer.

I wrote this example too quickly this morning and the `foreach `is obviously outside the try/catch : answer edited.

And you are absoluty right about the error messages : nice guess ! 

I found the parameter in the `RoundRobinTransport` class you told about
https://github.com/symfony/mailer/blob/d89611a7830d51b5e118bca38e390dea92f9ea06/Transport/RoundRobinTransport.php#L38

And because `FailoverTransport` class extends `RoundRobinTransport` it explains why the problem is the same if I am using failover(...) instead of roundrobin(...) options.

Thank you for the explaination and opening https://github.com/symfony/symfony/issues/52551

**Reply by ghost** [reply]
Until this issue will be solved inside Symfony you can patch this class as follows: copy this class, commit to your repository in new directory and keep the namespace, change constructor's default value, add entry inside composer.json `autoload -> psr-4`, run `composer dumpautoload` and it will use your class, not the one from `vendor/`.

However don't forget about this patch when upgrading symfony/mailer because original RoundRobinTransport class can be changed between two versions so you patch should be updated as well.

**Comment by sdespont** [comment]
I confirm that sleeping for 61 seconds in the catch allows sending the next emails.

**Comment by sdespont** [comment]
FYI https://github.com/symfony/symfony/pull/54939
